### ISSUE TITLE
Documentation for how the lifecycle of before/each/all/act/it/after should work

### ISSUE THREAD

**Issue by petemounce** [mattflo/NSpec#79]
It's not clear to me how the lifecycle of nspec tests goes. It seems like it should be possible to mix class level and delegate level before/act/it/after, but it doesn't work like I expect it to. I expect

0. class level before_all
0. class level before_each
0. class level act_each
0. each method in a class with an underscore
  0. delegate level before_all, within the method
    0. delegate level before (what is the difference between before, beforeAll and beforeEach??)
    0. delegate level before_each
    0. delegate level act
    0. it assertions
    0. delegate level after_each
  0. delegate level after_all
0. class level after_each
0. class level after_all

My context is that I'm writing outside-in API tests using OWIN+TestServer. I want to have a base class to take care of IoC, AutoFixture, NLog (I assert against logs in a memory-target), IoC container and TestServer. I think I want:
* once-per-class
  * set up autofixture
  * set up logging
* once per test method
  * once per context
    * set up container & poke in fakes where I want (I only want to fake away at my process boundaries)
    * set up TestServer with container

I've chosen to use xunit v2 as my debuggershim rather than nunit. I've chosen to use fields to hold my AutoFixture, Container and TestServer instances, on the assumption that I get a new instance of my whole class per test-method (which might be a wrong assumption). I have helper methods to create each item. I need those items to be available so that I can issue http requests to the app, having first substituted in fakes configured the way I need.

I've tried to write a spec to illustrate the combinations, which is below. I'm not sure I'm seeing what I'm expecting, to be honest. It would be nice to see a canonical diagram illustrating the lifecycle!

Are class fields the right thing to use, here? The examples on the home page illustrate that, but I'm guessing maybe doing that is the root of my issue - because I don't understand the lifecycle and what happens when.

If I can't use class fields, then I'll need to create helper methods to create each thing and return an instance that I can then use within my tests. But then I'm back to a world of duplicating setups and I seem to have lost most of the benefits that nspec brings regarding expressing behaviour concisely. I'm convinced I'm missing an ah-hah moment...?

Is there an illustration somewhere of how to use an auto-mocking container with nspec and nested contexts...?

Could the before/act/etc methods and delegates take and return a dynamic, that is the mutated state that they are responsible for doing, that is then fed through the pipeline so I can use that to assert against at the end...?

```csharp
using XUnit;
using NLog;
using NLog.Targets;
using Shouldly;

    public class nspec_reminder_doc : nspec_base
    {
        public void when_greeting()
        {
//            it["should be ok"] = () => "ok".ShouldBe("ok");
//            it["should be silly"] = () => "silly".ShouldBe("silly");
        }

        public void before_all()
        {
            _output += "cl_before_all,";
        }

        public void before_each()
        {
            _output += "cl_before_each,";
        }

        public void act_each()
        {
            _output += "cl_act_each,";
        }

        public void after_each()
        {
            _output += "cl_after_each,";
        }

        public void after_all()
        {
            _output += "cl_after_all,";
        }

        private string _output = "";

        public void remind_lifecycle_class_level()
        {
            after = () => _output = "";
            it["class level should print"] = () =>
                _output.ShouldBe("cl_before_all,cl_before_each,cl_act_each,");
            it["should print 2ndly because it unwraps and then wraps"] = () =>
                _output.ShouldBe("cl_before_all,cl_before_each,cl_act_each,");
        }

        public void remind_lifecycle_both_class_level_and_delegate()
        {
            beforeAll = () => _output += "del_beforeAll,";
            beforeEach = () => _output += "del_beforeEach,";
            before = () => _output += "del_before,";
            it["mix class level and delegate level should print"] = () =>
                _output.ShouldBe("cl_before_all,cl_before_each,del_beforeAll,del_beforeEach,");
            it["2nd time mix class level and delegate level should print"] = () =>
                _output.ShouldBe("cl_before_all,cl_before_each,del_beforeAll,del_beforeEach,");
        }

        public void remind_lifecycle_with_context()
        {
            beforeAll = () => _output += "del_beforeAll,";
            beforeEach = () => _output += "del_beforeEach,";
            before = () => _output += "del_before,";
            context["1"] = () =>
            {
                beforeAll = () => _output += "1_del_beforeAll,";
                beforeEach = () => _output += "1_del_beforeEach,";
                before = () => _output += "1_del_before,";
                it["context 1 should print"] = () =>
                    _output.ShouldBe("cl_before_all,cl_before_each,cl_act_each,");
            };
            context["2"] = () =>
            {
                beforeAll = () => _output += "2_del_beforeAll,";
                beforeEach = () => _output += "2_del_beforeEach,";
                before = () => _output += "2_del_before,";
                it["context 2 should print"] = () =>
                    _output.ShouldBe("cl_before_all,cl_before_each,cl_act_each,");
            };
        }
    }

    public abstract class nspec_base : nspec
    {
        [Fact]
        public void run()
        {
            var currentSpec = GetType();
            var finder = new SpecFinder(new[] {currentSpec});
            var builder = new ContextBuilder(finder, new Tags().Parse(currentSpec.Name), new DefaultConventions());
            var runner = new ContextRunner(builder, new ConsoleFormatter(), false);
            var collection = builder.Contexts().Build();
            var results = runner.Run(collection);

            //assert that there aren't any failures
            var failures = results.Failures();
            var count = failures.Count();
            if (count > 0)
            {
                var target = LogManager.Configuration.FindTargetByName("memory") as MemoryTarget;
                target.Logs.Do(Console.WriteLine);
            }
            count.ShouldBe(0);

            // TODO: figure out how to programmatically skip, when there are pending
        }
    }
}
```

**Comment by mattflo** [comment]
Admission: I haven't yet read through all of the context you shared. Take a look through the specs in this folder and see if they help you figure out the order and frequency of setup/tear down bits: 

https://github.com/mattflo/NSpec/tree/master/NSpecSpecs/describe_RunningSpecs/describe_before_and_after

Let me know of that helps.

**Comment by amirrajan** [comment]
Wrote this a while back: https://github.com/mattflo/NSpec/wiki/Execution-Orders

**Comment by BennieCopeland** [comment]
I'm noticing the same issue. To me it appears that the higher level beforeEach's are ran outside of their originating context. When nesting contexts, I expect for the current context to build upon the previous one. What I expect is that a nested contexts beforeAll will run AFTER it's parent's beforeEach. For example, I have an acceptance test that tests retrieving users from a database.

```
class User_Management_Feature : nspec
{
    void before_all()
    {
        // do something
    }

    void before_each()
    {
        // Start in memory HttpServer
        // Drop and Create DB
        // Wireup HttpClient
    }

    void Getting_a_list_of_users()
    {
        context["Given several users are registered"] = () =>
        {
            beforeAll = () =>
            {
                AddUserToDB(User1);
                AddUserToDB(User2);
                AddUserToDB(User3);
            }  // Fiddle the DB to add users to DB

            context["When I get the list of users"] = () =>
            {
                HttpResponseMessage response = null;
                act = () => // Send an HTTP GET request and save response

                it["I receive an HTTP 200 OK response"] = () => // Check response.StatusCode

                it["I am returned the list of users"] = () => // Check response.Content
            };
        };
    }
}
```

The order I expected is
1. Class level **beforeAll**
2. Class level **beforeEach**
3. Given several users are registered **beforeAll**
4. When I get the list of users **act**
5. Check response **it**
6. When I get the list of users **act**
7. Check list **it**

The order I get is
1. Class level **beforeAll**
2. Given several users are registered **beforeAll**
3. Class level **beforeEach**
4. When I get the list of users **act**
5. Check response **it**
6. Class level **beforeEach**
7. When I get the list of users **act**
8. Check list **it**

The **Given several users are registered** beforeAll runs slow and I would prefer to only run it once for all the tests. But because the **Class level** beforeEach runs not only after the beforeAll, but twice as well, it ends up dropping and recreating the database AFTER I populated it.

In my opinion, the order of execution should be:
1. beforeAll - Runs once for this context before everything else
2. beforeEach - Runs once for each **It**, and direct child **Context**
3. act - Runs once for each **It** and direct child **Context**
4. it - Runs
5. afterEach - Runs once for each **Act**, **it**, and direct child **Context**
6. afterAll - Runs once for this context after everything else

Using a bit of pseudo code:
```
void ExecuteContext(Context context)
{
    context.BeforeAll();
    foreach(var it in context.Its)
    {
        context.BeforeEach();
        context.Act();
        it();
        context.AfterEach();
    }

    foreach(var sub in context.SubContexts)
    {
        context.BeforeEach();
        context.Act();
        ExecuteContext(sub);
        context.AfterEach();
    }
    context.AfterAll();
}
```

**Comment by BennieCopeland** [comment]
Looking at https://github.com/mattflo/NSpec/wiki/Execution-Orders it appears that the beforeAll/afterAll are pinned to the **context** they were defined in, where as the beforeEach/afterEach are pinned to each **it**.

**Comment by BrainCrumbz** [comment]
I think it's the same thing that has bitten us in the past. The way one can think of it is that a `beforeAll` block found inside a context is executed when that context is *met* for the first time, and is executed only once no matter how many specifications are within the context. 

Then, after `beforeAll` runs, all specifications come into play and, as you said, any `before` block (or `beforeEach`, that's just an alias) is run according to enumerated `it` specs.

If that same context also has an `afterAll` block then, after all specifications ran, that block will run only once, before *leaving* the context.

**Comment by BennieCopeland** [comment]
I'm trying to think of any reason for why I would use `beforeAll` the way it is now. It runs counter intuitive to the way you would think it would with how it appears in the code. It's now harder to reason about the order of the setup. I viewed it as starting from the very basics of what is needed and building up an execution context by scopes. I viewed `context` as the same as `it`. For example, if I had:
```csharp
void do_something()
{
    beforeEach = () => { /* do some setup */ };

    it["should do too much"] = () =>
    { /* execute several actions and several expectations */ };
}
```

If I felt that the `it` was growing too big and doing too much, I might want to break it up into a `context`:
```csharp
void do_something()
{
    beforeEach = () => { /* do some setup */ };

    context["do action 1"] = () =>
    {
        beforeAll = () => { /* do a shared setup */ };

        it["should expect action1"] = () => { /* do something and check */ };

        context["and then do action 2"] = () =>
        {
            it["should expect action2"] = () => { /* do something and check */ };
        };
    };
    it["should expect something"] = () => { // execute action and check several expectations };
}
```

You would think that everything executes top to bottom the way it is viewed in the code, but it does not. It moves all the `beforeAll`'s to the top, and all the `beforeEach`'s to the bottom of the execution stack.

**Comment by BrainCrumbz** [comment]
As one can see from this [nested_contexts](https://github.com/mattflo/NSpec/blob/master/NSpecSpecs/describe_RunningSpecs/describe_before_and_after/nested_contexts.cs) test, yes, `beforeAll` hooks run before any `before` hooks or, as you say, they're moved to the top of execution stack. The opposite happens w.r.t to `after` hooks and `afterAll` hooks.

I'd suggest a different way of tackling specifications and of extending them when needed, because viewing `context` the same as (or a bigger brother than) `it` can be misleading. 

The way I suggest is: 

1. To think of `context` as a *scenario* where to verify your expectations, almost in a *use case* meaning. 
1. Within the context, setup everything needed to prepare that scenario by means of `before`, which would basically implement the *Given* part of a Given-When-Then specification.
1. Implement the main *action* triggering the test in the one `act` block, basically mapping it to the *When* part.
1. For that scenario, assert everything that should hold true with separate `it` blocks (<=> *Then* part)
1. Still within the context, cleanup as needed in an `after` block.

We're finding this way to write specs effective. At the beginning we were *abusing* somehow `before` hooks, including main action in them as well, but then we found ourselves in troubles when we wanted to refactor and grow tests. In that case, *grow* meant to add more sub-scenarios to a main scenario (i.e. add sub-contexts to a context), with different setup but with the same triggering action.

HTH

**Comment by BrainCrumbz** [comment]
Although I must say that, looking at sequence in https://github.com/mattflo/NSpec/wiki/Execution-Orders, it seems that nested `beforeAll`s are executed when their context is *met*, and they're not all moved up to the top:

```
...
common: before all 
child: before all  
...
nested: before all 
...
```

**Comment by BennieCopeland** [comment]
It's probably more accurate of me to say that the `beforeEach`'s are all moved after the `beforeAll`. So a `common: before each` and `child: before each` that are defined in higher `context`s than the `nested: before all`, and logically (at least in my mind) should run before `nested: before all`, are instead ran after. If we take that example and break it out into tree:
```
common: before all
    child: before all
        common: before each
            child: before each
                it works 1
            child: after each
        common: after each
        common: before each
            child: before each
                it works 2
            child: after each
        common: after each
        nested: before all
            common: before each
                child: before each
                    nested: before each
                        it works 3
                    nested: after each
                child: after each
            common: after each
            common: before each
                child: before each
                    nested: before each
                        it works 4
                    nested: after each
                child: after each
            common: after each
        nested: after all
    child: after all
common: after all
```

and we break out the direct path to `it works 4` it builds and destroys the context like this:
```
common: before all
    child: before all
        nested: before all
            common: before each
                child: before each
                    nested: before each
                        it works 4
                    nested: after each
                child: after each
            common: after each
        nested: after all
    child: after all
common: after all
```

The execution path that I was envisioning based on how it appears in code and how I thought `beforeAll` acted is:
```
common: before all
    common: before each
        child: before all
            child: before each
                it works 1
            child: after each
            child: before each
                it works 2
            child: after each
            child: before each
                nested: before all
                    nested: before each
                        it works 3
                    nested: after each
                    nested: before each
                        it works 4
                    nested: after each
                nested: after all
            child: after each
        child: after all
    common: after each
common: after all
```

This builds and destroys the context using this path:
```
common: before all
    common: before each
        child: before all
            child: before each
                nested: before all
                    nested: before each
                        it works 4
                    nested: after each
                nested: after all
            child: after each
        child: after all
    common: after each
common: after all
```

**Comment by BrainCrumbz** [comment]
BTW it seems this reasoning has already happened before: https://github.com/mattflo/NSpec/issues/51

Don't know what was the outcome at that time, maybe @amirrajan can shed a light. 

Our 2 cents: stick with whatever is RSpec "standard" behavior, or Mocha, so to reduce surprises for people coming from other dev languages

**Comment by amirrajan** [comment]
+1 on:

>Stick with whatever is RSpec "standard" behavior

more specifically: if the same test run in RSpec yields different results, then NSpec has a bug

**Comment by amirrajan** [comment]
Close? /cc @BrainCrumbz

**Comment by BrainCrumbz** [comment]
It still has recent activity, maybe @BennieCopeland still has feedbacks or queries. If we see no activity in the close future, we might as well close this.

BTW, at some point we mentioned on explicitly referencing this *order thing* in documentation, better if in main readme. What do you think about inserting a link to [Execution-Orders](https://github.com/mattflo/NSpec/wiki/Execution-Orders) wiki?

**Comment by amirrajan** [comment]
That and probably worth adding something to the website.

**Comment by BrainCrumbz** [comment]
Ok, started a [doc/execution-order branch](https://github.com/mattflo/NSpec/tree/doc/execution-order) for the first purpose.

**Comment by BrainCrumbz** [comment]
Please have a look at [this gist](https://gist.github.com/BrainCrumbz/5fde579cee880eeb9822). That's how we intend to modify the code block and output block in [Execution-Orders](https://github.com/mattflo/NSpec/wiki/Execution-Orders) wiki page.

I'm using the Gist as I think that wiki page commits do not go on the branch, so before applying that we'd like to know what do you think about it.

We were also thinking about adding that test case to source code, as an additional test case under SampleSpecs\Demo namespace

**Comment by BrainCrumbz** [comment]
Up

**Comment by BennieCopeland** [comment]
I have no further comments about changing the functionality. I do offer an alternative [gist](https://gist.github.com/BennieCopeland/cb9f2b27c81dd6a795ec) for the code and output blocks though.

On the topic of documentation, a page about making your test project a console app and using the built in "focus" tag would have been welcome when I was starting out. I found the mention of making it a console app from a tweet from you @amirrajan, and found the "focus" tag from digging through the source code for this issue.

**Comment by amirrajan** [comment]
Thanks @BennieCopeland, I'll be sure to add that in the new website stuff.

**Comment by BrainCrumbz** [comment]
@BennieCopeland goog alternative. On the visual side, I guess it's more about what gets a better picture in one's mind, and also where do you want to put the *stress* on. 

I liked the idea to show what gets executed from which context-level, so that at some point one sees an *unexpected* `method: before all` out-of-flow, so to say.  

Nonetheless you proposal is good, I'll take that for the wiki page.

**Comment by BennieCopeland** [comment]
@BrainCrumbz I added a little blurb on the wiki to explicitly point out the RSpec convention. Now that you explained about the out of flow `method: before all`, the output of your gist makes more sense. Looking at the output of my gist, the `beforeAll`s don't really jump out at you. What do you think of changing the code to this output? I think it better illustrates where the `beforeAll`s happen.

```
parent: before all
   child: before all
      parent: before each
      child: before each
      parent: it works 1
      child: after each
      parent: after each

      parent: before each
      child: before each
      parent: it works 2
      child: after each
      parent: after each

      parent: before each
      child: before each
      child: it works 3
      child: after each
      parent: after each

      parent: before each
      child: before each
      child: it works 4
      child: after each
      parent: after each

      method: before all
         parent: before each
         child: before each
         method: before each
         method: it works 5
         method: after each
         child: after each
         parent: after each

         parent: before each
         child: before each
         method: before each
         method: it works 6
         method: after each
         child: after each
         parent: after each

         sub: before all
            parent: before each
            child: before each
            method: before each
            sub: before each
            sub: it works 7
            sub: after each
            method: after each
            child: after each
            parent: after each

            parent: before each
            child: before each
            method: before each
            sub: before each
            sub: it works 8
            sub: after each
            method: after each
            child: after each
            parent: after each

         sub: after all
      method: after all
   child: after all
parent: after all
```

**Comment by BrainCrumbz** [comment]
Coool! Go for it. When you're done, I'll update the sample spec as well in [describe_execution_order.cs
](https://github.com/mattflo/NSpec/blob/doc/execution-order/SampleSpecs/Demo/describe_execution_order.cs) for PR #103

**Comment by BennieCopeland** [comment]
@BrainCrumbz The Wiki and my gist is updated with the new code and output.

**Comment by BrainCrumbz** [comment]
Thanks. PR updated.

**Comment by BennieCopeland** [comment]
@BrainCrumbz I think the code might need to be tweaked again. I took off the `abstract` from the parent class and the output changes to:
```
parent: before all
   parent: before each
   parent: it works 1
   parent: after each

   parent: before each
   parent: it works 2
   parent: after each

   child: before all
      parent: before each
      child: before each
      child: it works 3
      child: after each
      parent: after each

      parent: before each
      child: before each
      child: it works 4
      child: after each
      parent: after each

      ...
```
Maybe I'll add an `abstract class grandparent_class` and have `class parent_class` inherit from it to show that an abstract class's `it`s will be executed in the derived class's context, whereas an inherited concrete base class's `it`s will be executed in the base class. I'll work on getting something sensible together for you.

**Comment by BrainCrumbz** [comment]
I see. Maybe when printing `grandparent_class` messages use a `abstract:` prefix (instead of a more generic `grandparent:` one) to stress the difference with *simple* inheritance

**Comment by BennieCopeland** [comment]
That's what I'm thinking. I'm also thinking about breaking the page up into two sections. One focused on `beforeAll/beforeEach` execution order, then the next on the effects of inheritance.

**Comment by BennieCopeland** [comment]
I'm not sure if this is a bug or not, but when inheriting from a concrete class, the concrete class's `before/after`s are not ran in the same contexts.

```
    class concreteparent_class : nspec
    {
        public const int indentSize = 3;
        public string order = "\n\n";
        public int indent = 0;
        public string blah = "";

        protected void Print(string s)
        {
            Console.WriteLine(s);
        }

        protected void Increase(string s)
        {
            Write(s);
            indent += indentSize;
        }

        protected void Decrease(string s)
        {
            indent -= indentSize;
            Write(s);
        }

        protected void Write(string s)
        {
            order += s.PadLeft(s.Length + indent);
        }

        void before_all()
        {
            Increase("concreteparent: before all\n");
            blah = "Hello";
        }

        void before_each()
        {
            Write("concreteparent: before each\n");
        }

        void it_works_child_3()
        {
            Write("concreteparent: it works 3\n");
        }

        void it_works_child_4()
        {
            Write("concreteparent: it works 4\n");
        }

        void after_each()
        {
            Write("concreteparent: after each\n\n");
        }

        void after_all()
        {
            Decrease("concreteparent: after all\n\n");
            Print(order);
        }
    }

    class child_from_concrete_class : concreteparent_class
    {
        void before_all()
        {
            Increase("child: before all\n");
            blah.should_be("Hello");
        }

        void before_each()
        {
            Write("child: before each\n");
        }

        void it_works_child_5()
        {
            Write("child: it works 5\n");
        }

        void it_works_child_6()
        {
            Write("child: it works 6\n");
        }

        void after_each()
        {
            Write("child: after each\n");
        }

        void after_all()
        {
            Decrease("child: after all\n");
            Print(order);
        }
        void method_level_context()
        {
            beforeAll = () => Increase("method: before all\n");

            before = () => Write("method: before each\n");

            it["it works method 7"] = () => Write("method: it works 7\n");

            it["it works method 8"] = () => Write("method: it works 8\n");

            after = () => Write("method: after each\n");

            afterAll = () => Decrease("method: after all\n");

            context["sub context"] = () =>
            {
                beforeAll = () => Increase("sub: before all\n");

                before = () => Write("sub: before each\n");

                it["it works sub 9"] = () => Write("sub: it works 9 \n");

                it["it works sub 10"] = () => Write("sub: it works 10 \n");

                after = () => Write("sub: after each\n");

                afterAll = () => Decrease("sub: after all\n");
            };
        }
    }
```

The runner output shows as if they are nested, but the string output shows they are not. Additionally, a string set in the base class `beforeAll` and checked in the derived class shows as empty.

```
concreteparent class
  it works child 3
  it works child 4
  child from concrete class
    it works child 5
    it works child 6
    method level context
      it works method 7
      it works method 8
      sub context
        it works sub 9
        it works sub 10


child: before all
   concreteparent: before each
   child: before each
   child: it works 5
   child: after each
   concreteparent: after each

   concreteparent: before each
   child: before each
   child: it works 6
   child: after each
   concreteparent: after each

   method: before all
      concreteparent: before each
      child: before each
      method: before each
      method: it works 7
      method: after each
      child: after each
      concreteparent: after each

      concreteparent: before each
      child: before each
      method: before each
      method: it works 8
      method: after each
      child: after each
      concreteparent: after each

      sub: before all
         concreteparent: before each
         child: before each
         method: before each
         sub: before each
         sub: it works 9
         sub: after each
         method: after each
         child: after each
         concreteparent: after each

         concreteparent: before each
         child: before each
         method: before each
         sub: before each
         sub: it works 10
         sub: after each
         method: after each
         child: after each
         concreteparent: after each

      sub: after all
   method: after all
child: after all



concreteparent: before all
   concreteparent: before each
   concreteparent: it works 3
   concreteparent: after each

   concreteparent: before each
   concreteparent: it works 4
   concreteparent: after each

concreteparent: after all
```

**Comment by BrainCrumbz** [comment]
I suggest to go like this: for time being, let's leave the wiki page with the abstract/concrete inheritance scenario and with the sub-context scenario. Whether to keep those two in the same example, I don't know, up to you (for me it'd be ok in the same sample).

Then, open a new issue to clarify also with @amirrajan how inheritance is supposed to work, both with an abstract and a concrete parent. After that clarification (and possible a fix if behaviour is wrong), we'll eventually update docs with further (and clearer) scenarios.

What's your feedback.

**Comment by BennieCopeland** [comment]
Sounds good to me.

**Comment by alarya** [comment]
> Wrote this a while back: https://github.com/mattflo/NSpec/wiki/Execution-Orders

This only worked for me when I made the parent class **abstract**
